### ISSUE TITLE
Simplify / reorganize

### ISSUE THREAD

**Issue by gaborcsardi** [r-lib/async#33]
- [x] Remove `await()` (Our `await()` is essentially a synchronization barrier, and not needed.)
- [x] Make HTTP functions lazy
- [x] Make `async_constant()` lazy
- [x] Make it impossible to `resolve()` and `reject()` from the `action()`. Or rather, make all deferred  values lazy.
- [x] Remove `$get_result()` (Only one function needs the result synchronously, `synchronise()`, which will get it via internal methods.)
- [x] Remove `get_state()`, not needed.
- [x] ~~Rewrite `$catch()` to allow specifying multiple handlers. This applies to a promise chain, so it will be a member function, after all.~~
- [x] Remove the `on_rejected()` callback argument. We just need to use `$catch()` to handle errors.
- [x] Make it a run-time error if a promise does not have a `$then` follower in the chain,  and no `synchronise()` is waiting on it. No more computation paths that just lead to nowhere!
- [x] Make sure that errors bubble up to `synchronise()`. I.e. every time `null()` is used the error is still propagated, upwards.
- [x] Keep a reference to the parent promises.
- [x] Keep references to child promises, not just to the callback  functions. (This is needed for proper cancellation.)
- [x] On rejection, check if this would cause the parent promise(s) to go nowhere. If yes, then cancel those promises.
- [x] Allow multiple parents for a deferred value.
- [x] Get rid of the `null()` calls.
- [x] Allow changing the DAG of async computation dynamically. This is needed when a deferred evaluates to another deferred, or for joins like `when_all()` and `when_any()`.
- [x] Test cancellation for collection methods.

**Comment by gaborcsardi** [comment]
`$catch()` will be in another PR.
